include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp

```diff
@@ -37,998 +30,301 @@ namespace alpaka
-                return ::atomicInc(addr, value);
-            }
-        };
-        //! The GPU CUDA/HIPaccelerator atomic operation.
-        template<typename THierarchy>
-        struct AtomicOp<AtomicInc, AtomicUniformCudaHipBuiltIn, unsigned long int, THierarchy>
-        {
+                    return *reinterpretAddress(&value);
+                }
+            };
+            //! Emulate atomic
             //
-            __device__ static auto atomicOp(
-                AtomicUniformCudaHipBuiltIn const&,
-                [[maybe_unused]] unsigned long int* const addr,
-                [[maybe_unused]] unsigned long int const& value) -> unsigned long int
-            {
-                // LLP64
-                if constexpr(std::numeric_limits<unsigned int>::max() == std::numeric_limits<unsigned long>::max())
-                    return ::atomicInc(reinterpret_cast<unsigned int*>(addr), static_cast<unsigned int>(value));
-                else // LP64
+            // The default implementation will emulate all atomic functions with atomicCAS.
+            template<
+                typename TOp,
+                typename TAtomic,
+                typename T,
+                typename THierarchy,
+                typename TSfinae = void,
+                typename TDefer = void>
+            struct EmulateAtomic : private EmulationBase
+            {
+            public:
+                __device__ static auto atomic(
+                    alpaka::AtomicUniformCudaHipBuiltIn const& ctx,
+                    T* const addr,
+                    T const& value) -> T
                 {
-                    unsigned long long int* address_as_u = reinterpret_cast<unsigned long long int*>(addr);
-                    unsigned long long int old(*address_as_u);
-                    unsigned long long int assumed, eval;
+                    // Code from: http://docs.nvidia.com/cuda/cuda-c-programming-guide/#atomic-functions
+                    auto* const addressAsIntegralType = reinterpretAddress(addr);
+                    using EmulatedType = ALPAKA_DECAY_T(decltype(*addressAsIntegralType));
+                    EmulatedType old = *addressAsIntegralType;
+                    EmulatedType assumed;
                     do
                     {
                         assumed = old;
-                        eval = assumed >= static_cast<unsigned long long int>(value) ? 0 : assumed + 1;
-                        old = ::atomicCAS(address_as_u, assumed, eval);
+                        T v = *(reinterpret_cast<T*>(&assumed));
+                        TOp{}(&v, value);
+                        using Cas = alpaka::trait::
+                            AtomicOp<alpaka::AtomicCas, alpaka::AtomicUniformCudaHipBuiltIn, EmulatedType, THierarchy>;
+                        old = Cas::atomicOp(ctx, addressAsIntegralType, assumed, reinterpretValue(v));
+                        // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
```

Comment:
Do we have this case covered in tests? It should have been even before this PR, but since you are working on this part

Reply:
No, we do not cover this case.

Reply:
Our checks for atomicDec also do not test the flip over.